sda/overlay.go

```diff
@@ -30,20 +31,135 @@ type Overlay struct {
 	instancesLock     sync.Mutex
 	protocolInstances map[TokenID]ProtocolInstance
 
+	// treeMarshal that needs to be converted to Tree but host does not have the
+	// entityList associated yet.
+	// map from Roster.ID => trees that use this entity list
+	pendingTreeMarshal map[RosterID][]*TreeMarshal
+	// lock associated with pending TreeMarshal
+	pendingTreeLock sync.Mutex
+
+	// pendingSDAData are a list of message we received that does not correspond
+	// to any local Tree or/and Roster. We first request theses so we can
+	// instantiate properly protocolInstance that will use these SDAData msg.
+	pendingSDAs []*ProtocolMsg
+	// lock associated with pending SDAdata
+	pendingSDAsLock sync.Mutex
+
 	transmitMux sync.Mutex
 }
 
 // NewOverlay creates a new overlay-structure
 func NewOverlay(h *Host) *Overlay {
-	return &Overlay{
-		host:              h,
-		trees:             make(map[TreeID]*Tree),
-		entityLists:       make(map[RosterID]*Roster),
-		cache:             NewTreeNodeCache(),
-		instances:         make(map[TokenID]*TreeNodeInstance),
-		instancesInfo:     make(map[TokenID]bool),
-		protocolInstances: make(map[TokenID]ProtocolInstance),
-		transmitMux:       sync.Mutex{},
+	o := &Overlay{
+		host:               h,
+		trees:              make(map[TreeID]*Tree),
+		entityLists:        make(map[RosterID]*Roster),
+		cache:              NewTreeNodeCache(),
+		instances:          make(map[TokenID]*TreeNodeInstance),
+		instancesInfo:      make(map[TokenID]bool),
+		protocolInstances:  make(map[TokenID]ProtocolInstance),
+		transmitMux:        sync.Mutex{},
+		pendingTreeMarshal: make(map[RosterID][]*TreeMarshal),
+		pendingSDAs:        make([]*ProtocolMsg, 0),
+	}
+	// messages going to protocol instances
+	h.RegisterProcessor(o, SDADataMessageID)
+	// requesting a tree
+	h.RegisterProcessor(o, RequestTreeMessageID)
+	// sending back a tree
+	h.RegisterProcessor(o, SendTreeMessageID)
+	// requesting a roster
+	h.RegisterProcessor(o, RequestRosterMessageID)
+	// sending back a roster
+	h.RegisterProcessor(o, SendRosterMessageID)
```

Comment:
`types` would be the function-type? Yes, looks good. I wanted to do the same for the service-messages and other handlers, too.
Perhaps put it in plural like `RegisterProcessors`?